nova_checker/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+use codespan_reporting::diagnostic::{Diagnostic, Label};
+use hashbrown::{hash_set, HashMap, HashSet};
+use nova_parser::ast::{BinaryOp, Binding, BindingLevel, Expr, Stmt, UnaryOp};
+use std::{cell::RefCell, fmt::Display, hash::BuildHasher, rc::Rc};
+
+#[derive(Debug)]
+pub struct Env<'a> {
+    pub parent: Option<Rc<RefCell<Env<'a>>>>,
+    pub entries: HashMap<&'a str, Type>,
+}
+
+impl<'a> Env<'a> {
+    pub fn new() -> Self {
+        Self {
+            parent: None,
+            entries: HashMap::new(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Checker<'a> {
+    pub source: &'a str,
+    pub diagnostics: Vec<Diagnostic<usize>>,
+}
+
+#[derive(Debug, Clone)]
+pub enum Error {
+    InvalidBinaryOperator {
+        expr: Expr,
+        lhs_t: Type,
+        rhs_t: Type,
+    },
+}
+
+impl Error {
+    pub fn into_diagnostic(&self, fid: usize) -> Diagnostic<usize> {
+        match self {
+            Self::InvalidBinaryOperator {
+                expr:
+                    Expr::BinaryOp {
+                        op_index,
+                        kind,
+                        lhs,
+                        rhs,
+                    },
+                lhs_t,
+                rhs_t,
+            } => Diagnostic::error()
+                .with_code("E2365")
+                .with_message(format!(
+                    "Operator '{}' cannot be applied to types '{}' and '{}'.",
+                    kind.as_ref(),
+                    lhs_t,
+                    rhs_t,
+                ))
+                .with_labels(vec![
+                    Label::primary(fid, *op_index as usize..(*op_index as usize + 1)),
+                    Label::secondary(fid, lhs.span()).with_message(format!("{lhs_t}")),
+                    Label::secondary(fid, rhs.span()).with_message(format!("{rhs_t}")),
+                ]),
+
+            _ => panic!(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Type {
+    Undefined,
+    Number,
+    Boolean,
+    String,
+    Null,
+    Union(HashSet<Type>),
+}
+
+impl std::fmt::Display for Type {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Undefined => write!(f, "undefined"),
+            Self::Number => write!(f, "number"),
+            Self::Boolean => write!(f, "boolean"),
+            Self::String => write!(f, "string"),
+            Self::Null => write!(f, "null"),
+            _ => panic!(),
+        }
```

Comment:
Implement a `to_string` or something of that sort to `Type` instead, then `write!(f, self.to_str())`?